src/ellswift.rs

```diff
@@ -0,0 +1,686 @@
+    ///     let es = ElligatorSwift::from_seckey(&secp, sk, None);
+    /// # }
+    /// ```
+    pub fn from_seckey<C: Context>(
+        secp: &Secp256k1<C>,
+        sk: SecretKey,
+        aux_rand: Option<[u8; 32]>,
+    ) -> ElligatorSwift {
+        let mut es_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        let aux_rand_ptr =
+            aux_rand.as_ref().map(|rand| rand.as_c_ptr()).unwrap_or(core::ptr::null());
+        unsafe {
+            secp256k1_ellswift_create(
+                secp.ctx().as_ptr(),
+                es_out.as_mut_c_ptr(),
+                sk.as_c_ptr(),
+                aux_rand_ptr,
+            );
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(es_out))
+    }
+    /// Computes the ElligatorSwift encoding for a valid public key
+    /// # Example
+    /// ```
+    /// # #[cfg(feature = "alloc")] {
+    ///     use secp256k1::{ellswift::ElligatorSwift, PublicKey, Secp256k1, SecretKey};
+    ///     let secp = Secp256k1::new();
+    ///     let sk = SecretKey::from_slice(&[1; 32]).unwrap();
+    ///     let pk = PublicKey::from_secret_key(&secp, &sk);
+    ///     let es = ElligatorSwift::from_pubkey(pk);
+    /// # }
+    ///
+    /// ```
+    pub fn from_pubkey(pk: PublicKey) -> ElligatorSwift { Self::ellswift_encode(pk) }
+    /// Computes a shared secret only known by Alice and Bob. This is obtained by computing
+    /// the x-only Elliptic Curve Diffie-Hellman (ECDH) shared secret between Alice and Bob.
+    /// # Example
+    /// ```
+    /// # #[cfg(feature = "alloc")] {
+    ///     use secp256k1::{
+    ///         ellswift::{ElligatorSwift, ElligatorSwiftParty},
+    ///         PublicKey, SecretKey, XOnlyPublicKey, Secp256k1,
+    ///     };
+    ///     use core::str::FromStr;
+    ///
+    ///     let secp = Secp256k1::new();
+    ///
+    ///     let alice_sk = SecretKey::from_str("e714e76bdd67ad9f495683c37934148f4efc25ce3f01652c8a906498339e1f3a").unwrap();
+    ///     let bob_sk = SecretKey::from_str("b6c4b0e2f8c4359caf356a618cd1649d18790a1d67f7c2d1e4760e04c785db4f").unwrap();
+    ///
+    ///     let alice_es = ElligatorSwift::from_seckey(&secp, alice_sk, None);
+    ///     let bob_es = ElligatorSwift::from_seckey(&secp, bob_sk, None);
+    ///
+    ///     let alice_shared_secret = ElligatorSwift::shared_secret(alice_es, bob_es, alice_sk, ElligatorSwiftParty::A, None);
+    ///     let bob_shared_secret = ElligatorSwift::shared_secret(alice_es, bob_es, bob_sk, ElligatorSwiftParty::B, None);
+    ///
+    ///     assert_eq!(alice_shared_secret, bob_shared_secret);
+    /// # }
+    /// ```
+    pub fn shared_secret(
+        ellswift_a: ElligatorSwift,
+        ellswift_b: ElligatorSwift,
+        secret_key: SecretKey,
+        party: ElligatorSwiftParty,
+        data: Option<&[u8]>,
+    ) -> ElligatorSwiftSharedSecret {
+        let mut shared_secret = [0u8; 32];
+        unsafe {
+            secp256k1_ellswift_xdh(
```

Comment:
Is there a reason that the return value of this function call is not checked?

Reply:
We should `debug_assert` it. It can fail if one of the `ElligatorSwift` objects is invalid (prevented at the type level) or if `hashfp` fails (which it can't because `hash_callback` never returns 0).

Reply:
Added a `debug_assert_eq` in all calls of C code.